File: python/problem_083.py
# -*- coding: utf-8 -*-
"""
Problem 83 - Path sum: four ways

NOTE: This problem is a significantly more challenging version of Problem 81.

In the 5 by 5 matrix below, the minimal path sum from the top left to the
bottom right, by moving left, right, up, and down, is indicated using
parentheses and is equal to 2297.

                   (131)   673   (234)  (103) ( 18)
                   (201)  ( 96)  (342)   965  (150)
                    630    803   (746)  (422) (111)
                    537    699    497   (121)  956
                    805    732    524   ( 37) (331)

Find the minimal path sum, in "../resources/p083_matrix.txt", a 31K text file
containing a 80 by 80 matrix, from the top left to the bottom right by moving
left, right, up, and down.
"""


class Cell(object):

    def __init__(self, cost, best, predecessor):
        self.cost = cost
        self.best = best
        self.pred = predecessor

    def update(self, other):
        """
        Returns `True` if a move from `other` to `self` results in a better
        path
        """
        if self.best > other.best + self.cost:
            self.best = other.best + self.cost
            self.pred = other
            return True

        return False


def solution():
    matrix = []

    with open('../resources/p082_matrix.txt') as f:
        for line in f.readlines():
            matrix.append(map(int, line.strip().split(',')))

    maxrow = len(matrix) - 1
    maxcol = len(matrix[0]) - 1

    # Upper bound on the path sum
    maxpath = sum(cost for row in matrix for cost in row)

    cells = {(i, j): Cell(cost=matrix[i][j], best=maxpath, predecessor=None)
             for i in range(maxrow+1) for j in range(maxcol+1)}

    # Correct the best estimate for the first cell
    cells[0, 0].best = matrix[0][0]

    # Compute a first estimate of the best path to each cell by restricting
    # possible moves to right and down
    latest = [(0, 0)]

    while latest:
        new = []

        for row, col in latest:
            cell = cells[row, col]

            if row < maxrow and cells[row+1, col].update(cell):
                new.append((row+1, col))

            if col < maxcol and cells[row, col+1].update(cell):
                new.append((row, col+1))

        latest = new

    # Compute the real best path to each cell using continuous updates, until
    # the solution converges for all cells
    while True:
        updated = False

        for row, col in cells:
            cell = cells[row, col]

            # Iterate through all possible moves
            for (dx, dy) in ((-1, 0), (0, -1), (1, 0), (0, 1)):
                try:
                    updated |= cells[row+dx, col+dy].update(cell)
                except KeyError:
                    pass

        if not updated:
            break

    return cells[maxrow, maxcol].best


if __name__ == '__main__':
    print(solution())

File: go/problem_083.go
// Problem 83 - Path sum: four ways
//
// NOTE: This problem is a significantly more challenging version of Problem 81.
//
// In the 5 by 5 matrix below, the minimal path sum from the top left to the
// bottom right, by moving left, right, up, and down, is indicated using
// parentheses and is equal to 2297.
//
//                    (131)   673   (234)  (103) ( 18)
//                    (201)  ( 96)  (342)   965  (150)
//                     630    803   (746)  (422) (111)
//                     537    699    497   (121)  956
//                     805    732    524   ( 37) (331)
//
// Find the minimal path sum, in "../resources/p083_matrix.txt", a 31K text file
// containing a 80 by 80 matrix, from the top left to the bottom right by moving
// left, right, up, and down.
package main

import (
	"fmt"
	"io/ioutil"
	"strings"

	"./common"
)

type point struct {
	Row, Col int
}

type cell struct {
	Cost, Best int
}

// Returns `true` if a move from `other` to the current cell results in a more
// optimal path
func (c *cell) Update(other *cell) bool {
	if newCost := other.Best + c.Cost; newCost < c.Best {
		c.Best = newCost
		return true
	}
	return false
}

func main() {
	fmt.Println(solution())
}

func solution() int {
	bytes, _ := ioutil.ReadFile("../resources/p083_matrix.txt")
	contents := strings.TrimSpace(string(bytes))

	matrix := make([][]int, 0)

	for _, line := range strings.Split(contents, "\n") {
		matrix = append(matrix, common.Ints(strings.Split(line, ",")))
	}

	var maxPath int
	for _, rowValues := range matrix {
		maxPath += common.Sum(rowValues)
	}

	maxRow, maxCol := len(matrix)-1, len(matrix[0])-1

	cells := make(map[point]*cell)

	for row := 0; row <= maxRow; row++ {
		for col := 0; col <= maxCol; col++ {
			cells[point{row, col}] = &cell{Cost: matrix[row][col], Best: maxPath}
		}
	}

	cells[point{0, 0}].Best = matrix[0][0]

	// Compute a first estimate on the best path by by limiting the possible
	// moves to `Down` and `Right`
	paths := []point{point{0, 0}}

	for len(paths) > 0 {
		newPaths := make([]point, 0)

		for _, pt := range paths {
			cell := cells[pt]
			row, col := pt.Row, pt.Col

			if row < maxRow && cells[point{row + 1, col}].Update(cell) {
				newPaths = append(newPaths, point{row + 1, col})
			}

			if col < maxCol && cells[point{row, col + 1}].Update(cell) {
				newPaths = append(newPaths, point{row, col + 1})
			}
		}

		paths = newPaths
	}

	// Compute the real best path to each cell using continuous updates, until
	// the solution converges for all cells
	for {
		updated := false

		for pt, c := range cells {
			row, col := pt.Row, pt.Col

			neighbors := make([]*cell, 0)
			if row > 0 {
				neighbors = append(neighbors, cells[point{row - 1, col}])
			}
			if row < maxRow {
				neighbors = append(neighbors, cells[point{row + 1, col}])
			}
			if col > 0 {
				neighbors = append(neighbors, cells[point{row, col - 1}])
			}
			if col < maxCol {
				neighbors = append(neighbors, cells[point{row, col + 1}])
			}

			for _, neighbor := range neighbors {
				updated = updated || neighbor.Update(c)
			}
		}

		if !updated {
			break
		}
	}

	return cells[point{maxRow, maxCol}].Best
}
